Address Discovery and Advertisement

aarshkshah1992 opened the pull request:
@jacobheun Please can you review and approve ? Wanna get this up on the docs site.

jacobheun commented on content/concepts/address-discovery-and-advertisement.md:
```diff
@@ -0,0 +1,256 @@
+    - Creating long lived connections with the Relay server and advertising relay addresses in `AutoRelay`.
+
+    #### AutoNAT: Discovering we are NOT dialable/NATT’d
+    - To discover that we are NOT dialable, we use a component called [AutoNAT][autonat_repo].
```
That section might be a bit light, but I'd prefer to see the details here reflected there and then just link with a brief overview.

jacobheun commented on content/concepts/address-discovery-and-advertisement.md:
```diff
@@ -0,0 +1,256 @@
+
+    - Creating long lived connections with the Relay server and advertising relay addresses in `AutoRelay`.
+
+    #### AutoNAT: Discovering we are NOT dialable/NATT’d
```
Formatting looks like it's messed up on these headers due to the indenting. Should probably also be a level 5 header.

aarshkshah1992 replied:
Have revisited and fixed all headers.

jacobheun reviewed (commented):
Still need to review the last bit, but here are some suggestions. I'd recommend giving it another pass over after looking at the feedback. I can re review tomorrow/wednesday depending on when the updates land.

It's going to be great to have this more detailed info on the docs site ❤️

aarshkshah1992 commented on content/concepts/address-discovery-and-advertisement.md:
```diff
@@ -0,0 +1,256 @@
+
+The following address sets make up a libp2p node’s complete **set of dialable addresses**:
+
+- The network interface addresses a node explicitly listens on/binds to.
```
This is done.

aarshkshah1992 commented on content/concepts/address-discovery-and-advertisement.md:
```diff
@@ -0,0 +1,256 @@
+
+#### Publicly visible address(“Observed Addresses”) discovered using Identify
+
+- If the peer is behind a NAT, the NAT device would assign a NATT’d address to any outgoing packets sent by the peer.
```
Have addressed both of these valid comments.

aarshkshah1992 commented on content/concepts/address-discovery-and-advertisement.md:
```diff
@@ -0,0 +1,256 @@
+  NATT’d and are not dialable at all. The basic idea is that we connect to a publicly reachable “relay” server and
+  then create a “relay address” which we can advertise.
+  A relay address is usually a combination of the Relay server’s address combined with it’s peer id.
+  Any peer that then wants to talk to us can do so via the relay server.
```
This is done.

aarshkshah1992 commented:
@jacobheun 

Have addressed your review comments and have given another pass to the document. Have also updated the `AutoNAT` section on the `NAT traversal` page with more details. Please take a look.

jacobheun commented on content/concepts/address-discovery-and-advertisement.md:
```diff
@@ -0,0 +1,216 @@
+
+### Advertising Dialable Addresses
+
+The first section covered how a libp2p nodes “discovers” it’s dialable addresses.
+All the methods and address sets discussed above go into making the final set of a node’s dialable addresses.
+The way a libp2p peer shares these dialable addresses with other peers is:
```
```suggestion
The first section covered how a libp2p nodes "discovers" its dialable addresses. All the methods and address sets discussed above go into making the final set of a node’s advertised addresses. The way a libp2p peer shares these dialable addresses with other peers is as follows:
```

jacobheun commented on content/concepts/address-discovery-and-advertisement.md:
```diff
@@ -0,0 +1,216 @@
+All the methods and address sets discussed above go into making the final set of a node’s dialable addresses.
+The way a libp2p peer shares these dialable addresses with other peers is:
+
+- When a peer connects to other peers, it asks it’s `Host` for it’s current set of dialable addresses.
+
+- This will include some or all of the address sets discussed above (Interface, Relay, etc. etc.).
```
```suggestion
- When a peer connects to other peers, they exchange their dialable addresses via the `Identify` protocol.
    - This will include some or all of the address sets discussed above (Interface, Relay, etc. etc.).
```

jacobheun commented on content/concepts/address-discovery-and-advertisement.md:
```diff
@@ -0,0 +1,216 @@
+
+- This will include some or all of the address sets discussed above (Interface, Relay, etc. etc.).
+
+- Put those addresses in the `listenAddrs` field of the `Identify` message and send the `Identify` message to
+  the other peer.
+
```
This is confusing and is now covered by the previous point. I don't think we need to specify the protobuf field.
```suggestion
```

jacobheun commented on content/concepts/address-discovery-and-advertisement.md:
```diff
@@ -0,0 +1,216 @@
+- Put those addresses in the `listenAddrs` field of the `Identify` message and send the `Identify` message to
+  the other peer.
+
+- When a peer receives another peer’s dialable addresses from the `listenAddrs` field of the `Identify` message,
+  it stores them in it’s [peerstore][peerstore_repo] for some time. The timing does NOT matter for this discussion.
```
Just remove the timing mention since we're not talking about it.
```suggestion
- When a peer receives another peer’s dialable addresses from the `Identify` exchange it stores them in its [peerstore][peerstore_repo].
```

jacobheun commented on content/concepts/address-discovery-and-advertisement.md:
```diff
@@ -0,0 +1,216 @@
+- When a peer receives another peer’s dialable addresses from the `listenAddrs` field of the `Identify` message,
+  it stores them in it’s [peerstore][peerstore_repo] for some time. The timing does NOT matter for this discussion.
+
+- Any peer can then lookup a peer’s addresses by asking the DHT network to find the peer with that [peerID][peer_id] by
+  using the [FindPeer][find_peer] DHT API. The DHT nodes will use the addresses stored in their peerstore as described above
+  to find and return the addresses of the peer.
```
We don't mention how this actually succeeds as there is no mention of me connecting to my closest peers. The lookup won't work if I dont regularly connect to my closest peers.
```suggestion
- In order to allow other peers to discover our addresses, libp2p will intermittently query the DHT for its closest peers and connect to them. This will result in addresses being exchanged and stored with those peers, so that future DHT [FindPeer][find_peer] queries searching for our [peerID][peer_id] will discover our addresses.
```

jacobheun commented on content/concepts/address-discovery-and-advertisement.md:
```diff
@@ -0,0 +1,216 @@
+  using the [FindPeer][find_peer] DHT API. The DHT nodes will use the addresses stored in their peerstore as described above
+  to find and return the addresses of the peer.
+
+    #### Handling changes to the peer’s dialable addresses
```
```suggestion
#### Handling changes to the peer’s dialable addresses
```

jacobheun commented on content/concepts/address-discovery-and-advertisement.md:
```diff
@@ -0,0 +1,216 @@
+
+    #### Handling changes to the peer’s dialable addresses
+
+    All well and good so far. But the set of a peer’s dialable addresses is ofcourse not a static set. A peer can
+    start listening on new interfaces, see more of it’s observed addresses, change it’s network, be assigned a new
+    IP address by the Router, etc. Hence, we also need a mechanism to detect changes to a peer’s dialable addresses set
+    and advertise these changes to the world on the fly. The way we do this is by
+    **running an event loop** in the `Host`.
```
Avoid language like "of course". It may not be obvious to the reader (new to libp2p) and it can be demeaning in those instances.

```suggestion
All well and good so far, but the set of a peer’s dialable addresses is not a static set. A peer can start listening on new interfaces, see more of its observed addresses, change its network, be assigned a new IP address by the Router or ISP, and more. Hence, we also need a mechanism to detect changes to a peer’s dialable addresses set and advertise these changes to the world as they happen. In go-libp2p we do this is by **running an event loop** in the `Host`.
```

jacobheun commented on content/concepts/address-discovery-and-advertisement.md:
```diff
@@ -0,0 +1,216 @@
+    **running an event loop** in the `Host`.
+
+
+    #### When does the Host event loop fire ?
```
```suggestion
##### When does the Host event loop fire ?
```

jacobheun commented on content/concepts/address-discovery-and-advertisement.md:
```diff
@@ -0,0 +1,216 @@
+
+
+    #### When does the Host event loop fire ?
+    - Every 5 seconds.
+
+    - When the user starts listening on a new interface address/es by calling `Host.Network.Listen(addrs...)`.
+
+    - When `AutoRelay` detects a change in it’s reachability(private -> public or vice versa), creates a
+    connection with a new relay server or loses connection with an existing relay server since all of these would
+    cause a change in it’s relay addresses.
+
+    #### What does the Host do when the event loop fires ?
+    - Updates it’s primary network interface address by using the netroute library as discussed above
+    (if it has changed). This takes care of network changes/routers assigning new addresses via DHCP.
+
+    - Builds a new address set by using all the addresses discovered by the various mechanisms described in the
+    `Address Discovery` section and compares it with the previous address set.
+
+    - If the new address set is different from the previous address set, it fires an `EvtLocalAddressesUpdated` event.
+
+    - The ID service listens for the `EvtLocalAddressesUpdated` event and pushes the new set of addresses to all
+    peers it is connected to via the `Identify Push` protocol.
+    The ID service is the same service that also handles the `Identify` protocol described above.
```
```suggestion
- On interval, every 5 seconds.

- When the user starts listening on a new interface address/es by calling `Host.Network.Listen(addrs...)`.

- When `AutoRelay` detects a change in it’s reachability(private -> public or vice versa), creates a connection with a new relay server, or loses connection with an existing relay server since all of these would cause a change in its relay addresses.

##### What does the Host do when the event loop fires ?
- Updates its primary network interface address by using the netroute library as discussed above (if it has changed). This takes care of network changes/routers assigning new addresses via DHCP.

- Builds a new address set by using all the addresses discovered by the various mechanisms described in the `Address Discovery` section and compares it with the previous address set.

- If the new address set is different from the previous address set, it fires an `EvtLocalAddressesUpdated` event.
    - The ID service listens for the `EvtLocalAddressesUpdated` event and pushes the new set of addresses to all peers it is connected to via the `Identify Push` protocol. The ID service is the same service that also handles the `Identify` protocol described above.
```

jacobheun commented on content/concepts/address-discovery-and-advertisement.md:
```diff
@@ -0,0 +1,216 @@
+
+
+{{% notice "note" %}}
+While the blog post mentions that a peer shares it’s dialable addresses with other peers in the
+`Identify` and `Identify Push` messages, what it actually shares is a `Signed Peer Record`
+containing all the addresses and signed with it’s private key.
+That way, anyone who receives that address set can verify that it was announced by the intended peer.
+You can read more about how this works at https://github.com/libp2p/specs/pull/217.
```
blog post -> page. Link to the RFC itself.
```suggestion
While this page mentions that a peer shares its dialable addresses with other peers in the `Identify` and `Identify Push` messages, what it actually shares is a `Signed Peer Record` containing all the addresses and signed with its private key. That way, anyone who receives that address set can verify that it was indeed created by that peer.
You can read more about how this works in the [Routing Record RFC](https://github.com/libp2p/specs/blob/master/RFC/0003-routing-records.md).
```

jacobheun commented on content/concepts/nat.md:
```diff
@@ -53,6 +53,27 @@ While the [identify protocol][spec_identify] described above lets peers inform e
 A libp2p protocol called AutoNAT lets peers request dial-backs from peers providing the AutoNAT service.
+The basic idea as per the implementation in `go-libp2p` is as follows:
+
+- An AutoNAT client send it’s list of [dialable addresses][address_link] to an AutoNAT server and asks the AutoNAT server to
+  see if it can dial the client back on one of those addresses.
```
```suggestion
- An AutoNAT client sends its list of [dialable addresses][address_link] to an AutoNAT server and asks the AutoNAT server to see if it can dial the client back on one of those addresses.
```

jacobheun commented on content/concepts/nat.md:
```diff
@@ -53,6 +53,27 @@ While the [identify protocol][spec_identify] described above lets peers inform e
+- An AutoNAT client send it’s list of [dialable addresses][address_link] to an AutoNAT server and asks the AutoNAT server to
+  see if it can dial the client back on one of those addresses.
+
+- If the AutoNAT server is successful, it replies to the original request with a success message along with
+  the address it was able to successfully dial on.
```
```suggestion
- If the AutoNAT server is successful, it replies to the original request with a success message along with the address it was able to successfully dial on.
```

jacobheun commented on content/concepts/nat.md:
```diff
@@ -53,6 +53,27 @@ While the [identify protocol][spec_identify] described above lets peers inform e
+
+- If the AutoNAT server fails to dial back the node, it sends back a failure message to the client.
+
+- A peer can become an AutoNAT server by providing services on the “"/libp2p/autonat/1.0.0" protocol.
```
```suggestion
- A peer can become an AutoNAT server by providing services on the "/libp2p/autonat/1.0.0" protocol.
```

jacobheun commented on content/concepts/nat.md:
```diff
@@ -53,6 +53,27 @@ While the [identify protocol][spec_identify] described above lets peers inform e
+
+- A peer can become an AutoNAT server by providing services on the “"/libp2p/autonat/1.0.0" protocol.
+
+- Peers that want to discover their reachability are AutoNAT clients.
+  They periodically pick a random AutoNAT server they are connected to, send the server it’s complete list of
+  dialable addresses(by asking the Host) and ask for a dial back.
+
+- Once an AutoNAT client has enough votes on whether it is publicly dialable(Public Reachability) or
+  NOT(private Reachability), it emits an `EvtLocalReachabilityChanged` event with the reachability information.
+
+- Note that the Reachability status can always flip from Private to Public and vice versa as the peer discovers
+  new dialable addresses and gets periodic dial back responses from AutoNAT servers.
```
```suggestion
- Peers that want to discover their reachability are AutoNAT clients. They periodically pick a random AutoNAT server they are connected to, send the server it’s complete list of dialable addresses(by asking the Host) and ask for a dial back.

- Once an AutoNAT client has enough information on whether it is publicly dialable(Public Reachability) or NOT(private Reachability), it emits an `EvtLocalReachabilityChanged` event with the reachability information.

- Note that the Reachability status can always flip from Private to Public and vice versa as the peer discovers new dialable addresses and gets periodic dial back responses from AutoNAT servers.
```

jacobheun reviewed (commented):
Mostly just language and styling cleanup

DannyS03 commented:
Based on the content found here being partly included in other documents and the change in structure, and the stale nature of the document, we can close this. We can also consider taking from this document or generating a similar one that is implementation agnostic. Please reopen if otherwise.
